src/main/java/net/kyori/adventure/text/minimessage/transformation/TransformationType.java

```diff
@@ -37,50 +43,180 @@
     this.canParse = canParse;
-    this.parser = parser;
+    this.factory = (ctx, name, args) -> {
+      final T ret = parser.parse();
+      ret.context(ctx);
+      ret.load(name, args);
+      return ret;
+    };
+  }
+
+  /**
+   * Constructs a new transformation type.
+   *
+   * @param canParse the predicate used to check if a tag can be parsed by this type
+   * @param factory the factory that should be used to create this type
+   * @since 4.2.0
+   */
+  private TransformationType(final Predicate<String> canParse, final TransformationFactory<T> factory) {
+    this.canParse = canParse;
+    this.factory = factory;
+  }
+
+  /**
+   * Create a new transformation type with dynamically determined names.
+   *
+   * <p>It is assumed that the {@code nameMatcher} function is side-effect free, meaning that any time
+   * it is called for a certain input {@code x}, it will always return the same value, and not modify any other state.</p>
+   *
+   * @param nameMatcher the name matcher predicate
+   * @param factory a factory
+   * @param <T> transformation instance type
+   * @return a new transformation type definition
+   * @since 4.2.0
+   */
+  public static <T extends Transformation> TransformationType<T> transformationType(final Predicate<String> nameMatcher, final TransformationFactory<T> factory) {
+    return new TransformationType<>(
+      requireNonNull(nameMatcher, "nameMatcher"),
+      requireNonNull(factory, "factory")
+    );
+  }
+
+  /**
+   * Create a new transformation type that can be parsed without performing recursive parses.
+   *
+   * @param nameMatcher filter for tag names to apply this transformation to
+   * @param contextFreeFactory the context-free factory
+   * @param <T> transformation instance type
+   * @return a new transformation type definition
+   * @see #transformationType(Predicate, TransformationFactory)
+   * @since 4.2.0
+   */
+  public static <T extends Transformation> TransformationType<T> transformationType(final Predicate<String> nameMatcher, final TransformationFactory.ContextFree<T> contextFreeFactory) {
+    return new TransformationType<>(
+      requireNonNull(nameMatcher, "nameMatcher"),
+      requireNonNull(contextFreeFactory, "contextFreeFactory")
+    );
+  }
+
+  /**
+   * Create a name matcher function that will accept the provided case-insensitive tag names.
+   *
+   * @param elements the accepted names
+   * @return a name matcher function
+   * @since 4.2.0
+   */
+  public static Predicate<String> acceptingNames(final String... elements) {
+    return acceptingNames(Arrays.asList(elements));
+  }
+
+  /**
+   * Create a name matcher function that will accept the provided case-insensitive tag names.
+   *
+   * @param elements the accepted names
+   * @return a name matcher function
+   * @since 4.2.0
+   */
+  public static Predicate<String> acceptingNames(final Collection<String> elements) {
```

Comment:
```suggestion
  public static Predicate<String> acceptingCaseInsensitiveNames(final Collection<String> elements) {
```

Reply:
hmm, i'm getting more and more tempted to force-lowercase every name passed into the predicate, which would effectively make every tag case-insensitive. Thoughts? (this would resolve #141)

Reply:
sounds good (but should be documented :D)